### ISSUE TITLE
when I run egs2/librimix/tse1/run.sh, the loss=0.000e+00 all the time

### ISSUE THREAD

**Issue by mubingshen** [espnet/espnet#5065]
I want to train a Target Speaker Extraction model on Librimix dataset, but I found the snr_loss and final loss(equal to snr_loss) are always 0.000e+00. Here is my train log:
node7:0/6] 2023-03-27 16:10:37,131 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:37,656 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:38,126 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:38,600 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:39,061 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:39,532 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:40,001 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:40,474 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:40,949 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:10:41,431 (trainer:663) WARNING: The grad norm is nan. Skipping updating the model.
[node7:0/6] 2023-03-27 16:11:32,619 (trainer:721) INFO: 1epoch:train:1-100batch: iter_time=0.095, forward_time=0.138, **snr_loss=8.605, loss=8.605**, backward_time=0.291, optim0_lr0=0.001, train_time=0.696, optim_step_time=0.082
[node7:0/6] 2023-03-27 16:12:30,770 (trainer:721) INFO: 1epoch:train:101-200batch: iter_time=0.006, forward_time=0.098, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.289, optim0_lr0=0.001, train_time=0.582, optim_step_time=0.078
[node7:0/6] 2023-03-27 16:13:28,737 (trainer:721) INFO: 1epoch:train:201-300batch: iter_time=0.008, forward_time=0.099, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.290, optim0_lr0=0.001, train_time=0.579, optim_step_time=0.079
[node7:0/6] 2023-03-27 16:14:27,015 (trainer:721) INFO: 1epoch:train:301-400batch: iter_time=0.007, forward_time=0.098, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.290, optim0_lr0=0.001, train_time=0.582, optim_step_time=0.079
[node7:0/6] 2023-03-27 16:15:26,911 (trainer:721) INFO: 1epoch:train:401-500batch: iter_time=0.008, forward_time=0.100, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.289, optim0_lr0=0.001, train_time=0.598, optim_step_time=0.079
[node7:0/6] 2023-03-27 16:16:28,703 (trainer:721) INFO: 1epoch:train:501-600batch: iter_time=0.006, forward_time=0.102, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.289, optim0_lr0=0.001, train_time=0.617, optim_step_time=0.078
[node7:0/6] 2023-03-27 16:17:27,648 (trainer:721) INFO: 1epoch:train:601-700batch: iter_time=0.009, forward_time=0.102, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.289, optim0_lr0=0.001, train_time=0.589, optim_step_time=0.082
[node7:0/6] 2023-03-27 16:18:27,095 (trainer:721) INFO: 1epoch:train:701-800batch: iter_time=0.006, forward_time=0.102, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.289, optim0_lr0=0.001, train_time=0.594, optim_step_time=0.082
[node7:0/6] 2023-03-27 16:19:24,407 (trainer:721) INFO: 1epoch:train:801-900batch: iter_time=0.008, forward_time=0.101, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.290, optim0_lr0=0.001, train_time=0.573, optim_step_time=0.082
[node7:0/6] 2023-03-27 16:20:24,664 (trainer:721) INFO: 1epoch:train:901-1000batch: iter_time=0.009, forward_time=0.103, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.290, optim0_lr0=0.001, train_time=0.602, optim_step_time=0.082
[node7:0/6] 2023-03-27 16:21:22,982 (trainer:721) INFO: 1epoch:train:1001-1100batch: iter_time=0.006, forward_time=0.101, **snr_loss=0.000e+00, loss=0.000e+00**, backward_time=0.290, optim0_lr0=0.001, train_time=0.583, optim_step_time=0.082

I think the loss has an anomaly, but I just run according to the configuration in the original yaml file(train_enh_tse_td_speakerbeam_4gpu_max.yaml). Who can help me analyze whether it is reasonable, thanks a lot!

**Comment by sw005320** [comment]
I just added @Emrys365 for this thread.

Most likely, something happened in the training stage since "The grad norm is nan."
I'm expecting that 
1. the optimization parameters are wrong
2. the data preparation is wrong

If you did not change anything from the recipe, it might be due to 1.
In this case, I'm curious whether you modify the training config or not.

**Comment by mubingshen** [comment]
Thank you for your reply!
In response to the two possible issues you mentioned, the optimization parameters are same as the original yaml file(train_enh_tse_td_speakerbeam_4gpu_max.yaml), which the optimization parameters are as follows:
_optim_conf:
    lr: 1.0e-03
    eps: 1.0e-08
    weight_decay: 0_
and I completed the process of data preparation according to the run.sh file.
Here is my train.yaml file:
optim: adam
max_epoch: 400
batch_type: folded
batch_size: 12
iterator_type: chunk
chunk_length: 24000
# num_workers: 4
optim_conf:
    lr: 1.0e-03
    eps: 1.0e-08
    weight_decay: 0
unused_parameters: true
allow_variable_data_keys: true
patience: 1
accum_grad: 1
grad_clip: 5.0
val_scheduler_criterion:
- valid
- loss
best_model_criterion:
-   - valid
    - si_snr
    - max
-   - valid
    - loss
    - min
keep_nbest_models: 1
scheduler: reducelronplateau
scheduler_conf:
   mode: min
   factor: 0.7
   patience: 3

model_conf:
    num_spk: 2
    share_encoder: true

train_spk2enroll: data/train/spk2enroll.json
enroll_segment: 24000
load_spk_embedding: false
load_all_speakers: false

encoder: conv
encoder_conf:
    channel: 256
    kernel_size: 16
    stride: 8
decoder: conv
decoder_conf:
    channel: 256
    kernel_size: 16
    stride: 8
extractor: td_speakerbeam
extractor_conf:
    layer: 8
    stack: 4
    bottleneck_dim: 256
    hidden_dim: 512
    skip_dim: 256
    kernel: 3
    causal: False
    norm_type: gLN
    nonlinear: relu
    # enrollment related
    i_adapt_layer: 7
    adapt_layer_type: mul
    adapt_enroll_dim: 256

# A list for criterions
# The overlall loss in the multi-task learning will be:
# loss = weight_1 * loss_1 + ... + weight_N * loss_N
# The default `weight` for each sub-loss is 1.0
criterions: 
  # The first criterion
  - name: snr 
    conf:
      eps: 1.0e-7
    wrapper: fixed_order
    wrapper_conf:
      weight: 1.0

**Comment by Emrys365** [comment]
Are you using the max version of Librimix for training? In this case, the chunk-based iteration should be used with caution because it might produce all-zero reference signals in some local region of a long sample where the target speaker is inactive.

**Comment by mubingshen** [comment]
Yes, I used the max version of Librimix for training. In your opinion, it is better for me to use min version of Librimix for training?

**Comment by mubingshen** [comment]
What's more, do I need to set the param **load_all_speakers=true**?

**Comment by mubingshen** [comment]
I printed the inf and ref tensor in snrloss function, and I found the inf tensor is always a all-zero tensor when the snrloss is zero. Sometimes the ref tensor is a all-zero tensor and the reason you mentioned above. I think the key problem is to know why the inf tensor is always a all-zero tensor.

**Comment by Emrys365** [comment]
Yes. I find it a bit messy to handle the all-zero reference issue in the training code. There's no simple solution to it. So we use the min version of the data by default.

A possible workaround is to use the default iterator instead of the chunk-based one. The disadvantage is higher memory consumption due to longer samples.

**Comment by Emrys365** [comment]
> What's more, do I need to set the param load_all_speakers=true?

Yes, you could do that to calculate the loss for all speakers in each mixture sample during training.

**Comment by mubingshen** [comment]
Thank you for your reply!
As I said above, inf tensor is a all-zero tensor when snrloss=0. I checked code line by line and found in **espnet2/enh/layers/tcn.py** **line244**, the **score** tensor is 
```
tensor([[[-18.1603, -19.1000, -18.3435,  ..., -17.6619, -19.3610, -17.8926],
         [-41.5312, -50.3529, -48.9905,  ..., -44.1537, -48.6571, -36.7862],
         [-43.9848, -46.9822, -46.1305,  ..., -42.2977, -39.8586, -32.8187],
         ...,
         [-46.6670, -50.4166, -50.2755,  ..., -45.1846, -43.0759, -36.4117],
         [-23.0000, -28.1384, -29.9785,  ..., -28.0685, -30.7993, -25.5061],
         [-40.1088, -45.7707, -43.8833,  ..., -41.6056, -45.5847, -39.3304]]],
       device='cuda:0', grad_fn=<ConvolutionBackward0>)
```
and then **score** tensor will be used as input to the **relu** activation function, this is why the inf tensor becomes all-zero tensor. When I changed activation function to **sigmoid**, the snrloss is not 0

**Comment by Emrys365** [comment]
I made a PR (https://github.com/espnet/espnet/pull/5155) to update the above mentioned modifications.

**Comment by JinZr** [comment]
hi all, 

i'm also training the librimix enh1 recipe with the default conf (the conv-tasnet one) on a dataset i simulated lately. i encountered the same issue with `iterator_type: chunk`, the problem disappeared when i removed this line from the training conf. i'm wondering if there's any specific reason of it or there might be some problems with the simulated data. 

fyi, i'm currently on git commit `71dc9a3f8c8e1608667017d9911618ad1cc96f65`.

thanks!

best
jin

**Comment by Emrys365** [comment]
The issue is that when chunk iterator is used, you may get chunks that do not contain the target speaker, hence the training target becomes all-zero, which makes the SI-SNR loss unstable.

**Comment by JinZr** [comment]
i see, so in this case would `load_all_speakers` help?

**Comment by Emrys365** [comment]
> A possible workaround is to use the default iterator instead of the chunk-based one. The disadvantage is higher memory consumption due to longer samples.

One simple solution is my reply earlier in this thread ⬆️

You can further apply the enh preprocessor in the latest code base with the following arguments to enable one-segment chunking for each sample while avoiding all-zero reference signals:
```yaml
preprocessor: enh
preprocessor_conf:
   speech_segment: 48000
   avoid_allzero_segment: true
```
